File: Anurag_19_CSE/Python/day30.py
# Link For Problem: https://leetcode.com/problems/divide-two-integers/

class Solution:

    def divide(self, dividend: int, divisor: int) -> int:
        is_negative = (dividend < 0) != (divisor < 0)

        divisor, dividend = abs(divisor), abs(dividend)

        quotient = 0
        the_sum = divisor

        while the_sum <= dividend:
            current_quotient = 1

            while (the_sum << 1) <= dividend:
                the_sum <<= 1
                current_quotient <<= 1

            dividend -= the_sum
            the_sum = divisor
            quotient += current_quotient

        return min(2147483647, max(-quotient if is_negative else quotient, -2147483648))

File: Anurag_19_CSE/Java/day30.java
//Link For Problem: https://leetcode.com/problems/divide-two-integers/

public class day30 {

    public int divide(int dividend, int divisor) {
        if (dividend == Integer.MIN_VALUE && divisor == -1)
            return Integer.MAX_VALUE;

        boolean negative = dividend < 0 ^ divisor < 0;

        dividend = Math.abs(dividend);
        divisor = Math.abs(divisor);

        int quotient = 0, subQuot = 0;

        while (dividend - divisor >= 0) {
            for (subQuot = 0; dividend - (divisor << subQuot << 1) >= 0; subQuot++)
                ;
            quotient += 1 << subQuot;
            dividend -= divisor << subQuot;
        }

        return negative ? -quotient : quotient;
    }
}
